### DISCUSSION TITLE
Proposal for bringing SnowFS to Bevy

### DISCUSSION THREAD

**Discussion by sebastianrath** [bevyengine/bevy#2484] [Ideas]
Hello, I am the creator of an open-source project called [SnowFS](https://github.com/snowtrack/snowfs). SnowFS is a lightweight and performant VCS with a focus on graphic projects and binary files.

The project is developed in TypeScript because it was initially developed for an Electron app. Due to my C++ roots, I planned to port SnowFS back to C++, but I am more and more eye-balling with Rust these days.

I heard there is a demand for a VCS in Bevy, and I would love to hear your thoughts on the following proposal.  Would you be interested in helping out bringing SnowFS to Rust / Bevy? If you think there is a potential fit, I would be happy to hear your thoughts on this either here or on Discord.

Any ideas or feedbacks on this topic is highly appreciated!

**Comment by alice-i-cecile** [comment]
Cool! Yes, version control integration is very much needed for both scenes and assets: Git is not really up to the task.

@cart could you convert this to a discussion? We can hash out the basic plan there (and in #editor on Discord) and then if there's buy-in I'd be happy to help out with the RFC process.

**Reply by IceSentry** [reply]
If scenes use a text based format I'm not sure why git wouldn't work for them.

**Reply by lukors** [reply]
@alice-i-cecile why is Git not up to the task?

**Comment by IceSentry** [comment]
I'm not sure how this relates to bevy. I understand it could be used by people making games using bevy, but does bevy need some kind of specific integration with the VCS to work properly with it?

Also, if I still want to host the code on github, could I just gitignore the assets folder and use SnowFS only for the assets folder or should I use it for the entire project.

**Reply by kirawi** [reply]
It's covered here: https://github.com/snowtrack/snowfs#gitgit-lfs
I'm not a game developer and I've only been passively watching Bevy, but from what I understand its important for the future editor. A VCS targeted towards binary file types is important for art because of how huge they are and how radically they change in a short timespan; the algorithms that Git uses just aren't suited for that.

**Reply by IceSentry** [reply]
No, I get it's advantages compared to git, what I don't get is why or if it needs a specific integration in bevy. A VCS is generally an external tool that doesn't require changing any code inside a project.

My point about using git is that I would like to still be able to use git for the ability to host the code on github while hosting the assets that are not text somewhere else using SnowFS. I was mostly asking if it has a way to ignore some files I guess.

**Reply by kirawi** [reply]
I'm not sure. However, I think Git has built-in mechanisms to allow you to use your own custom diff algorithm: `GIT_EXTERNAL_DIFF`. I'm guessing the intended way of doing this is not including the assets inside the git repository, and then keeping the assets with SnowFS somewhere else. I'm not an artist, so I don't know.

**Reply by sebastianrath** [reply]
> I was mostly asking if it has a way to ignore some files I guess.

Absolutely, you can easily exclude source code files through a file called `.snowignore`

**Reply by ManevilleF** [reply]
But why an integration in bevy? That's still an external tool

**Reply by kirawi** [reply]
To use for the editor.

**Comment by lukors** [comment]
My understanding after reading a bit about SnowFS is that it's exclusively local version control, is this correct?

Version control is nice, but it feels like it's not nearly as useful if it can only be used in single-person projects.

Maybe I'm misunderstanding something? Maybe you could outline some use cases?

**Reply by IceSentry** [reply]
I think it's only local only during the alpha and it will eventually support remote just like git. Otherwise, yeah, that's a pretty big limitation

**Reply by sebastianrath** [reply]
It depends very much on what functionality you are looking for. I've been told that version control is generally of interest. A full-blown solution like Perforce or Git(Hub), but in better form, actually requires a lot more and might require as many resources as Bevy itself (if not more).

Eventually SnowFS will provide basic remote support, but from what I understand from your posts, I'm not sure such remote feature would solve the problem. Bevy's target audience is indeed very technical and could easily run and manage their own servers, but I don't think setting up a local "SnowFS server" can compete with the convenience of a cloud solution. And we haven't talked about collaboration support yet. Is there a wish list somewhere for what is needed for Bevy?
